### PR TITLE
RHCLOUD-45374 - MCP S2 - Add tool to check custom role permissions and what actions are …

### PR DESCRIPTION
**Pull request by EvanCasey13**
…allowed

## Link(s) to Jira
https://redhat.atlassian.net/browse/RHCLOUD-45374

## Description of Intent of Change(s)
  - Add check_role_permissions MCP tool for pre-flight analysis of custom role permissions before assigning to users
  - Support both V1 and V2 organizations with automatic version detection
  - Provide comprehensive permission analysis: grouping by application, wildcard expansion, verb coverage, and
  actionable recommendations
  - Only check custom roles (tenant-specific), not system/seeded roles from public tenant

  Features

  - Role lookup: Case-insensitive exact match with partial match suggestions
  - Permission analysis: Groups permissions by application, expands wildcards, identifies included/missing verbs
  - Coverage analysis: Detects read-only roles, write capability, wildcard usage, cross-application permissions
  - Recommendations: Actionable insights like "Role grants read-only access" or "Consider adding :delete if full CRUD is
   needed"
  - Available permissions: Optional flag to list permissions the role does NOT include per application
 
## Local Testing
How can the feature be exercised?

# Run V1 tool tests
pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPCheckRolePermissionsTests

# Run V2 tool tests
pipenv run tox -e py312-fast -- tests.management.test_mcp_views.MCPCheckRolePermissionsV2Tests

How can the bug be exploited and fix confirmed?
Is any special local setup required?

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [ ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  - [ ] if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

## Summary by Sourcery

Introduce an MCP tool to perform pre-flight analysis of custom role permissions for both V1 and V2 organizations, returning structured permission coverage and recommendations.

New Features:
- Add check_role_permissions MCP tool that analyzes a custom role’s permissions and summarizes coverage, wildcard usage, and recommendations.
- Support automatic routing between V1 and V2 orgs when checking custom roles’ permissions, including org_version metadata in responses.

Enhancements:
- Provide optional listing of available-but-not-granted permissions per application for a role to aid gap analysis.
- Centralize shared role-permission analysis logic for V1 and V2 roles into a common helper for consistent output and recommendations.

Tests:
- Add comprehensive MCP tests for check_role_permissions covering success, error handling, wildcard expansion, recommendations, and org version reporting for V1.
- Add V2-specific tests for check_role_permissions, including seeded-role exclusion, auth behavior, and org_version=v2 responses.
- Add edge-case tests for wildcard verbs/resources, multi-application roles, and verb inclusion/exclusion behavior in recommendations.

### CODE REVIEW

File: rbac/management/mcp_views.py [hunk 1]
```diff
@@ -978,6 +980,255 @@ def _get_role_v2(request: HttpRequest, role_uuid: str) -> str:
...
+
+
+def _check_role_permissions_v1(
+    tenant: Tenant,
+    role_name: str,
+    include_available_permissions: bool,
+) -> str:
+    """Pre-flight check for a V1 custom role."""
+    # Step 1: Find the custom role by name (case-insensitive exact match first, then partial)
+    role = Role.objects.filter(name__iexact=role_name, tenant=tenant).first()
```
Comment (lpichler) [inline rbac/management/mcp_views.py]: The tool description states \"This tool only checks custom roles (tenant-specific), not system/seeded roles.\" While the `tenant=tenant` filter effectively excludes most system roles (since they reside in the public tenant), consider adding an explicit `system=False` filter to match the stated contract:

```python
role = Role.objects.filter(name__iexact=role_name, tenant=tenant, system=False).first()
```

Same for the partial-match fallback query on line 1032.

This protects against edge cases where a system role might exist under a non-public tenant.

**Severity**: Low

File: rbac/management/mcp_views.py [hunk 2]
```diff
@@ -978,6 +980,255 @@ def _get_role_v2(request: HttpRequest, role_uuid: str) -> str:
...
+
+
+def _check_role_permissions_v2(
+    tenant: Tenant,
+    role_name: str,
+    include_available_permissions: bool,
+) -> str:
+    """Pre-flight check for a V2 custom role."""
+    # Step 1: Find the custom role by name (case-insensitive exact match first, then partial)
+    role = RoleV2.objects.filter(name__iexact=role_name, tenant=tenant).first()
```
Comment (lpichler) [inline rbac/management/mcp_views.py:1219]: Similar to the V1 path above — the tool description says it only checks custom roles, but this query doesn't filter by `type`. A seeded or platform role that happens to exist in the tenant's space could be returned.

Consider adding `type=RoleV2.Types.CUSTOM`:

```python
role = RoleV2.objects.filter(name__iexact=role_name, tenant=tenant, type=RoleV2.Types.CUSTOM).first()
```

Same for the partial-match fallback query on line 1077.

**Severity**: Low

**Review by lpichler — APPROVED** [review]
